internal/wasm/sdk/internal/wasm/bindings.go

```diff
@@ -4,93 +4,196 @@
+		wasmtime.NewFunc(store, wasmtime.NewFuncType([]*wasmtime.ValType{i32, i32}, []*wasmtime.ValType{i32}), dispatcher.Call).AsExtern(),
+		wasmtime.NewFunc(store, wasmtime.NewFuncType([]*wasmtime.ValType{i32, i32, i32}, []*wasmtime.ValType{i32}), dispatcher.Call).AsExtern(),
+		wasmtime.NewFunc(store, wasmtime.NewFuncType([]*wasmtime.ValType{i32, i32, i32, i32}, []*wasmtime.ValType{i32}), dispatcher.Call).AsExtern(),
+		wasmtime.NewFunc(store, wasmtime.NewFuncType([]*wasmtime.ValType{i32, i32, i32, i32, i32}, []*wasmtime.ValType{i32}), dispatcher.Call).AsExtern(),
+		wasmtime.NewFunc(store, wasmtime.NewFuncType([]*wasmtime.ValType{i32, i32, i32, i32, i32, i32}, []*wasmtime.ValType{i32}), dispatcher.Call).AsExtern(),
 	}
 
-	imports, err = imports.AppendFunction("opa_abort", opa_abort, C.opa_abort)
-	if err != nil {
-		return nil, err
+	return externs
+}
+
+func opaAbort(caller *wasmtime.Caller, args []wasmtime.Val) ([]wasmtime.Val, *wasmtime.Trap) {
+
+	data := caller.GetExport("memory").Memory().UnsafeData()[args[0].I32():]
+
+	n := bytes.IndexByte(data, 0)
+	if n == -1 {
+		panic("invalid abort argument")
 	}
 
-	imports, err = imports.AppendFunction("opa_builtin0", opa_builtin0, C.opa_builtin0)
-	if err != nil {
-		return nil, err
+	panic(abortError{message: string(data[:n])})
+}
+
+type builtinDispatcher struct {
+	ctx      *topdown.BuiltinContext
+	builtins map[int32]topdown.BuiltinFunc
+	result   *ast.Term
+}
+
+func newBuiltinDispatcher() *builtinDispatcher {
+	return &builtinDispatcher{}
+}
+
+func (d *builtinDispatcher) SetMap(m map[int32]topdown.BuiltinFunc) {
+	d.builtins = m
+}
+
+func (d *builtinDispatcher) Reset(ctx context.Context) {
+	d.ctx = &topdown.BuiltinContext{
+		Context:  ctx,
+		Cancel:   nil,
+		Runtime:  nil,
+		Time:     ast.NumberTerm(json.Number(strconv.FormatInt(time.Now().UnixNano(), 10))),
+		Metrics:  metrics.New(),
+		Cache:    make(builtins.Cache),
+		Location: nil,
+		Tracers:  nil,
+		QueryID:  0,
+		ParentID: 0,
 	}
 
-	imports, err = imports.AppendFunction("opa_builtin1", opa_builtin1, C.opa_builtin1)
-	if err != nil {
-		return nil, err
+}
+
+func (d *builtinDispatcher) Call(caller *wasmtime.Caller, args []wasmtime.Val) (result []wasmtime.Val, trap *wasmtime.Trap) {
+
+	if d.ctx == nil {
+		panic("unreachable: uninitialized built-in dispatcher context")
 	}
 
-	imports, err = imports.AppendFunction("opa_builtin2", opa_builtin2, C.opa_builtin2)
-	if err != nil {
-		return nil, err
+	if d.builtins == nil {
+		panic("unreachable: uninitialized built-in dispatcher index")
 	}
 
-	imports, err = imports.AppendFunction("opa_builtin3", opa_builtin3, C.opa_builtin3)
+	exports := getExports(caller)
+
+	var convertedArgs []*ast.Term
+
+	// first two args are the built-in identifier and context structure
+	for i := 2; i < len(args); i++ {
+
+		x, err := fromWasmValue(exports, args[i].I32())
+		if err != nil {
+			panic(builtinError{err: err})
```

Comment:
So here's something I hadn't fully appreciated in the go-wasmer days either: Why do we need to panic? Is it because the method called _from Wasm_ cannot return anything to indicate that something went wrong?

Reply:
Yes, currently the built-in functions cannot return an error to our generated wasm. There's no room for that in the interface--the built-in function returns an address to the output term or NULL to indicate undefined. Perhaps we could replace the panics with Traps but I haven't looked into that yet and since the panic/defer approach seems to work, I've kept it here. There's definitely room for improvement here.

Reply:
Thanks!